### PR TITLE
Add locale fallbacks

### PR DESCRIPTION
**Pull request by nbramblett**
Adds the ability for the pomsg.Bundle() to load in fallback locales when given a locale that is too specific. As a brief example, if `en_US.po` does not exist, `en.po` will be used instead.

Ran `go test` to ensure everything is up to snuff, and added some new tests to help ensure reliability.

### CODE REVIEW

File: soymsg/pomsg/pomsg.go [hunk 1]
```diff
@@ -26,19 +27,37 @@ type FileOpener interface {
...
+// canonical form.
 func Load(opener FileOpener, locales []string) (soymsg.Provider, error) {
 	var prov = provider{make(map[string]soymsg.Bundle)}
 	for _, locale := range locales {
 		r, err := opener.Open(locale)
 		if err != nil {
 			return nil, err
 		} else if r == nil {
-			continue
+			localeTag, err := language.Parse(locale)
```
Comment (mancusi) [inline soymsg/pomsg/pomsg.go:41]: I personally think it'd be clearer if this if statement was just another function called "handleFallback" or something.

File: soymsg/pomsg/pomsg.go [hunk 2]
```diff
@@ -92,7 +109,20 @@ func Dir(dirname string) (soymsg.Provider, error) {
 }
 
 func (p provider) Bundle(locale string) soymsg.Bundle {
```
Comment (robfig) [inline soymsg/pomsg/pomsg.go:111]: This locale is guaranteed to be one of the ones passed to Load, right? 

In that case, it might be better to keep the mapping of "requested locale" to "actual locale" in the provider, rather than calculating fallback in two different places?

File: soymsg/pomsg/pomsg.go (hunk 1 above)
Comment (nbramblett) [reply soymsg/pomsg/pomsg.go:41]: acking because I wound up removing this

File: soymsg/pomsg/pomsg.go (hunk 2 above)
Comment (nbramblett) [reply soymsg/pomsg/pomsg.go:111]: That makes sense, I think the bundler is the only thing that _truly_ cares about fallbacks, everything else is largely redundancies. Putting up a patch now.

File: soymsg/pomsg/fallback_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		name string
+		localeCode string
+		expectedTags []language.Tag
+	}{
+		{
+			name: "When given a generic locale code, no extra fallbacks are provided",
+			localeCode: "en",
+			expectedTags: []language.Tag{language.English},
+		},
+		{
```
Comment (robfig) [inline soymsg/pomsg/fallback_test.go:19]: Do you mind adding a couple more cases? For example, invalid tag, one that uses a script (chinese)

File: soymsg/pomsg/pomsg_test.go [hunk 4]
```diff
@@ -13,7 +13,7 @@ func TestPOBundle(t *testing.T) {
 		t.Error(err)
 		return
 	}
-	var bundle = pomsgs.Bundle("zz")
+	var bundle = pomsgs.Bundle("en")
```
Comment (robfig) [inline soymsg/pomsg/pomsg_test.go:16]: For my understanding, why were these test changes necessary?

File: soymsg/pomsg/pomsg_test.go [hunk 5]
```diff
@@ -143,3 +148,46 @@ func TestNewMessage(t *testing.T) {
 		}
 	}
 }
+
+func TestFallbackBundle(t *testing.T) {
```
Comment (robfig) [inline soymsg/pomsg/pomsg_test.go:152]: This is sort of a lot of code to copy/paste. Maybe a loop over "en", "en_UK" is preferable in the other test?

File: soymsg/pomsg/pomsg_test.go (hunk 4 above)
Comment (nbramblett) [reply soymsg/pomsg/pomsg_test.go:16]: the new x/text/language package will throw an error ("tag is well-formed but unknown") when it tries to parse a locale that isn't defined by ISO_3166-1. Rather than re-write everything in the test to be compatible, it was much easier to just use a legitimate language tag.

File: soymsg/pomsg/fallback_test.go (hunk 3 above)
Comment (nbramblett) [reply soymsg/pomsg/fallback_test.go:19]: Done

File: soymsg/pomsg/pomsg_test.go (hunk 5 above)
Comment (nbramblett) [reply soymsg/pomsg/pomsg_test.go:152]: Done

File: soymsg/pomsg/fallback_test.go (hunk 3 above)
Comment (robfig) [reply soymsg/pomsg/fallback_test.go:19]: I don't believe either of the ZZ or Zzzz branches are covered.. maybe one more case that covers them?

File: soymsg/pomsg/pomsg_test.go (hunk 4 above)
Comment (robfig) [reply soymsg/pomsg/pomsg_test.go]: Oh, I think I see what's happening. This test was correctly identifying a behavior change. It used to return an empty bundle when the locale was not present (a zero bundle is usable), but now it returns nil. I think that should be corrected.

Comment (nbramblett) [reply soymsg/pomsg/pomsg_test.go]: Figured out that "zz" was failing because I had included locale parsing in Load originally. With it now only being present in Bundle, zz is a valid .po file. It just won't be able to do any sort of fallback checking on a fake locale.

File: soymsg/pomsg/pomsg_test.go [hunk 6]
```diff
@@ -70,7 +78,7 @@ func TestPlural(t *testing.T) {
 		return
 	}
 
-	const locale = "zz"
+	const locale = "en"
```
Comment (robfig) [inline soymsg/pomsg/pomsg_test.go:81]: Is this change still required?

Comment (nbramblett) [reply soymsg/pomsg/pomsg_test.go:81]: Nope, should be ok to swap back. Fixing.

**Comment by robfig** [comment]
Thanks!
